design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,679 @@
+kind: Stack
+metadata:
+  name: CachingWebServiceStack
+spec:
+  templates:
+    redis.example.org/v1:
+      deployment: |
+        kind: Deployment
+        …
+          containers:
+          - name: redis-controller
+            image: example/redis-controller:{{.spec.redisVersion}}
+    cachingwebservice.example.org/v1:
+      cache: |
+        apiVersion: example.org/v1
+        kind: Redis
+        …
+        spec:
+          redisVersion: "5"
+      web: |
+        kind: Deployment
+        …
+          containers:
+          - name: nginx-controller
+            image: example/nginx-controller:{{.spec.nginxVersion}}
+--
+apiVersion: example.org/v1
+kind: CachingWebService
+metadata:
+  name: cacheme
+spec:
+  nginxVersion: "1.17.4"
+```
+
+In the example above, distinct templates are defined for the Redis CRD and the
+CachingWebService. CachingWebService includes two templates, one of which relies
+on the other.
+
+The `nginxVersion` is defined in the `spec` of the `CachingWebService` instance,
+while `redisVersion` has been declared in the `cache` template for
+`cachingwebservice.example.org/v1`. In each case, a template variable has been
+passed into a dependent template. One was user supplied, the other was template
+supplied.
+
+An additional form of composition and reuse is available by referencing Template
+Stack kinds within new Template Stacks. For example, `MegaWebsite` could include
+`CachingWebServer` in the `dependsOn` field of that Stack.
+
+#### Using Reconciled Values as Template Variables
+
+Kubernetes resources typically set their state within a `status` field, which is
+often available as a Status Sub-Resource. In fact, CRD resources specifically
+support scale and status sub-resources. Access to this sub-resource can be
+granted independently from the resource itself.
+
+The template capabilities available to a Template Stack should not be restricted
+for use within the `spec` or other resource creation features.
+
+Template Stack managed resources should be able to manage their status, as any
+controller managed resource would. Arguably, the most useful values to expose
+through this status would be derived from resources created with templates.
+
+While maintaining that the `spec.templates` field of a `Stack` can present the
+entire template body of CRD resources, let `templateStatus` present a template
+for a resource's `status` body.  These `templateStatus` will need matching keys
+to the `template`. To match a `templateStatus` to the appropriate `template`,
+the GVK and name of the template will be needed.
+
+```yaml
+…
+kind: Stack
+metadata:
+  name: StatusExampleStack
+spec:
+  template:
+    example.example.org/v1:
+      templateA: |
+        kind: Deployment
+        spec:
+        …
+  templateStatus:
+    example.example.org/v1:
+      templateA: |
+        statusField: {{.some.value}}
+```
+
+*The `status` could be included in the previously described template body, but
```

Comment:
Your senses are correct about something being off here.

The `templateStatus` will not need to include the various template names.  There is a single `templateStatus` per CRD.

```
  templateStatus:
    example.example.org/v1:
      templateA: |
        statusField: {{.some.value}}
```

should be:

```
  templateStatus:
    example.example.org/v1: |
        statusField: {{.some.value}}
```

Reply:
> So if the author defines their CRD to use or not use a separate status subresource, we have to honor that. The TSM can't make it's own independent decision.

That's a good point.  The TSM can optimize to use `UpdateStatus` if a Status sub-resource is included in the CRD (which can be detected by the TSM), otherwise it will just update status with the rest of the resource.

Reply:
This less abstract example may help:

```yaml
apiVersion:  stacks.crossplane.io/v1alpha1
kind: Stack
metadata:
  name: Foo
spec:
   templates:
     foo.group/version:
       templateA:
         kind: athing
         spec:
           foovar: {{ .spec.foo }}
   templateStatus:
     foo.group/version:
       statusthing: {{ .templateA.status.bar }}
```

Given:

```yaml     
apiVersion:  stacks.crossplane.io/v1alpha1
kind: Foo
spec:
  foo: "foo"
```

The TSM will generate the following:

```yaml
kind: athing
spec:
  foovar: "foo"
```

As `athing` type is reconciled by an outside controller, a `status` emerges:

```yaml
kind: athing
spec:
  foovar: "foo"
status:
  bar: "bar"
```

On the next TSM pass, the Foo resource will be updated:

```yaml     
apiVersion:  stacks.crossplane.io/v1alpha1
kind: Foo
spec:
  foo: "foo"
status:
  statusthing: "bar"
```

Reply:
Perhaps this would be better?

```yaml
  templates:
     status:
     kindA.group/version:
       nameA: |
       nameB: |
     kindB.group/version:
       nameA: |
       nameB: |
````

Reply:
I'm leaving it up to https://github.com/crossplaneio/crossplane/pull/956 and related design updates to sort out how templates, status, and resources are defined.